### ISSUE TITLE
Consider switching to the Shift AST

### ISSUE THREAD

**Issue by sebmck** [babel/babel#858]
Things to consider:

 - This will need to be done before #607 if at all.
 - How to handle parsing. Right now we use [acorn-babel](https://github.com/babel/acorn-babel) that's a fork of [acorn-jsx](https://github.com/RReverser/acorn-jsx) that's a fork of [acorn](https://github.com/marijnh/acorn). Potential solutions:
   - Fork [shift-parser](https://github.com/shapesecurity/shift-parser-js) to include our custom extensions, playground features, ES7, flow, JSX etc.
   - Fork `acorn-babel` into `acorn-shift`. (Easiest and most viable)
 - How to handle regenerator since it uses SpiderMonkey. Potential solutions:
   - Deopt and convert the AST to SpiderMonkey to pass to regenerator if the AST contains any async/generator functions.
   - Fork/inline regenerator again. (Easiest and most viable)

/cc @RReverser

**Comment by eventualbuddha** [comment]
The actual shift parser was a fair bit slower than acorn last I checked. I used parsing jQuery's full source 1000 times as the benchmark and got something like 6s for acorn vs 9s for shift.js. Just something to consider. This was about a month ago, so things may have changed.

**Comment by thejameskyle** [comment]
ping @michaelficarra

**Comment by michaelficarra** [comment]
Closed? :cry: And we just finished ES6 support today, too. We'll be starting our two-phase parsing (which should be both more tolerant and faster) effort this week.

**Comment by sebmck** [comment]
@michaelficarra Haven't ruled it out! Main issue is the parser. I don't want any regressions so all the current parser feature set needs to be supported and I'm not sure how practical that is.

**Comment by sebmck** [comment]
@michaelficarra Any plans on making the shift parser "pluggable" similar to marijnh/acorn#168?

**Comment by michaelficarra** [comment]
Not currently. We'll be experimenting with a refactoring in esprima that should allow for extension. With the Shift parser, we're going to be implementing our two-phase parsing (grammar only, then early error reporting) in a way that does not allow extension. Depending on how that works for esprima, we might merge the two strategies for Shift.

**Comment by sebmck** [comment]
Reopening since I'm seriously considering this again. ESTree is starting to become really gross. I started making changes to ESTree to make it nicer but realised that I was actually making the same modifications as the Shift AST does. Wont be able to use the Shift parser but using the AST is definently viable.

**Comment by stefanpenner** [comment]
Also, re: some of the above performance concerns. Babel isn't currently limited by parser performance, at-least not in the cases that people typically associate with degraded performance. As @sebmck as been improving performance, one common trend becomes apparent, by reducing the number of AST traversals (and look-aheads and behinds) the performance improves nicely.

Based on some discussions with @michaelficarra and @ariya the shift-reducer is inline with the above observation.

I am by no means saying that the reducer is the silver bullet to our problems, practical or empirically  faster. Rather the approach may yield positives that outweigh differences in parser performance.

**Comment by sebmck** [comment]
IANAL but does anyone know if the Shift AST [LICENSE](https://github.com/shapesecurity/shift-spec/blob/es6/LICENSE) will be an issue? I have no idea what the licensing implications are of using it.

**Comment by michaelficarra** [comment]
@sebmck As I understand, Apache 2 is as permissive as BSD/MIT, with 2 differences:

0. it makes explicit all of the rights that are granted as opposed to the looser terms of basically "do what you want" in BSD/MIT (this makes it better in my opinion)
0. the reason our lawyer asked us to use this license: we retain trademark/patent rights

The license is both FSF and OSI approved. You can see how it compares favourably with other OSS licenses here: https://en.wikipedia.org/wiki/Comparison_of_free_and_open-source_software_licenses

@stefanpenner The Shift reducer is great, but if you find performance needs to be improved, know that we have plans to change its implementation from being directed by a representation of the spec at runtime to being generated from a representation of the spec during a compilation step.

**Comment by sebmck** [comment]
@michaelficarra Ah cool, just unsure about how it would effect the license that Babel is released under and what kind of notices would need to be included and where (if any).

**Comment by michaelficarra** [comment]
@sebmck Since you're not redistributing any of the projects released under the license (they're installed by npm afterward), nothing should need to change. If you were, you'd need to distribute a copy of our license with the software, just as you would with a BSD/MIT project.

**Comment by sebmck** [comment]
@michaelficarra Babel would just be switching to the Shift AST, meaning that the node definitions would be implemented independently in the Babel parser etc. The Shift parser and the other suite of packages wont necessarily be used.

**Comment by michaelficarra** [comment]
You should be fine, but the current Google/Oracle lawsuit over implementing the Java APIs comes to mind. I can confirm something with our lawyer on Monday if you have a specific question. But I'd say you already run a greater risk implementing the ESTree/SpiderMonkey API since no explicit license is given for it.

**Comment by RReverser** [comment]
@michaelficarra That's an interesting and valid concern. Maybe worth discussing in a separate issue under ESTree?

**Comment by sebmck** [comment]
@michaelficarra Sweet, a confirmation would be awesome. Thanks!

@RReverser Already is one, estree/estree#52.

**Comment by sebmck** [comment]
Closing this as it's not going to happen. I've played around with the Shift AST (even tried to convert parts of Babel to it) and I'm not convinced it makes anything easier, it actually makes converting nodes very awkward and tricky.
